autoload/maktaba/json.vim

```diff
@@ -0,0 +1,211 @@
+    " TODO(dbarnett): Escape other special characters.
+    return '"' . l:escaped . '"'
+  elseif maktaba#value#IsList(a:value)
+    let l:json_items = map(copy(a:value), 'maktaba#json#Format(v:val)')
+    return '[' . join(l:json_items, ', ') . ']'
+  elseif maktaba#value#IsDict(a:value)
+    if maktaba#value#IsCallable(a:value)
+      throw maktaba#error#BadValue(
+          \ 'Funcdict for func %s cannot be represented as JSON.',
+          \ string(a:value.func))
+    endif
+    let l:json_items =
+        \ map(items(a:value), 's:FormatKeyAndValue(v:val[0], v:val[1])')
+    return '{' . join(l:json_items, ', ') . '}'
+  endif
+  throw maktaba#error#BadValue(a:value)
+endfunction
+
+function! s:Consume(str, count) abort
+  return maktaba#string#StripLeading(a:str[a:count : ])
+endfunction
+
+function! s:ParsePartial(json, custom_values) abort
+  " null, true, or false
+  if a:json =~# '\m^null\>'
+    let l:value = get(a:custom_values, 'null', s:NULL)
+    return [l:value, s:Consume(a:json, 4)]
+  endif
+  if a:json =~# '\m^true\>'
+    let l:value = get(a:custom_values, 'true', s:TRUE)
+    return [l:value, s:Consume(a:json, 4)]
+  endif
+  if a:json =~# '\m^false\>'
+    let l:value = get(a:custom_values, 'false', s:FALSE)
+    return [l:value, s:Consume(a:json, 5)]
+  endif
+  " Number
+  " TODO(dbarnett): Handle scientific notation.
+  let l:num_match = matchstr(a:json, '\v^-?[0-9]+(\.[0-9]+)?>')
+  if !empty(l:num_match)
+    return [eval(l:num_match), s:Consume(a:json, len(l:num_match))]
+  endif
+  " String
+  let l:str_match = matchstr(a:json, '\v^"([^\\"]|\\.)*"')
+  if !empty(l:str_match)
+    " JSON strings use the same syntax as JSON strings.
+    " TODO(dbarnett): Handle special escape sequences like \uxxxx.
+    let l:value = eval(l:str_match)
+    return [l:value, s:Consume(a:json, len(l:str_match))]
+  endif
+  " First character if any, empty string otherwise.
+  let l:first_char = a:json[0:0]
+  " List
+  if l:first_char is# '['
+    let [l:items, l:remaining] = s:ParseListPartial(a:json, a:custom_values)
+    return [l:items, l:remaining]
+  endif
+  " Dict
+  if l:first_char is# '{'
+    let [l:dict, l:remaining] = s:ParseDictPartial(a:json, a:custom_values)
+    return [l:dict, l:remaining]
+  endif
+  throw maktaba#error#BadValue('Input is not valid JSON text.')
+endfunction
+
+function! s:ParseListPartial(json, custom_values) abort
+  let l:remaining = s:Consume(a:json, 1)
+  if l:remaining[0:0] is# ']'
+    return [[], s:Consume(l:remaining, 1)]
+  endif
+  let l:items = []
+  while !empty(l:remaining)
+    " Parse and consume one value as the next list item.
+    let [l:item, l:remaining] = s:ParsePartial(l:remaining, a:custom_values)
+    call add(l:items, l:item)
+    " If next character is "]", consume and finish list.
+    if l:remaining[0:0] is# ']'
+      let l:remaining = s:Consume(l:remaining, 1)
+      break
+    endif
+    " Only other acceptable character is comma. Consume and continue.
```

Comment:
Unless I messed up somewhere, every branch strips leading whitespace (in `s:Consume` and `s:ParsePartial`) so you can just check the first character and ignore whitespace on the rest of the parsing logic.